lib/jwt/jwt.go

```diff
@@ -0,0 +1,284 @@
+// key that form a *jwt.Key.
+type Config struct {
+	// Clock is used to control expiry time.
+	Clock clockwork.Clock
+
+	// PublicKey is a PEM encoded key that can be used to verify a signed token.
+	PublicKey []byte
+
+	// PrivateKey is a PEM encoded key that can be used to sign and verify tokens.
+	PrivateKey []byte
+}
+
+// CheckAndSetDefaults validates the values of a *Config.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.PrivateKey == nil && c.PublicKey == nil {
+		return trace.BadParameter("public or private key is required")
+	}
+	return nil
+}
+
+// Key is a JWT key that can be used to sign and/or verify a token.
+type Key struct {
+	*Config
+
+	publicKey  *rsa.PublicKey
+	privateKey *rsa.PrivateKey
+}
+
+// New creates a JWT key that can be used to sign and verify tokens.
+func New(config *Config) (*Key, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	privateKey, err := parsePrivateKey(config.PrivateKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	publicKey, ok := privateKey.Public().(*rsa.PublicKey)
+	if !ok {
+		return nil, trace.BadParameter("invalid public key")
+	}
+
+	return &Key{
+		Config:     config,
+		publicKey:  publicKey,
+		privateKey: privateKey,
+	}, nil
+}
+
+// NewPublic is a JWT key that can only be used to verify tokens.
+func NewPublic(config *Config) (*Key, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	publicKey, err := parsePublicKey(config.PublicKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Key{
+		Config:    config,
+		publicKey: publicKey,
+	}, nil
+}
+
+// SignParams are the claims to be embedded within the JWT token.
+type SignParams struct {
+	// Username is the Teleport identity.
+	Username string
+
+	// Roles are the roles assigned to the user within Teleport.
+	Roles []string
+
+	// Expiry is time to live for the token.
+	Expiry time.Duration
+}
+
+// Check verifies all the values are valid.
+func (p *SignParams) Check() error {
+	if p.Username == "" {
+		return trace.BadParameter("missing username")
+	}
+	if len(p.Roles) == 0 {
+		return trace.BadParameter("missing roles")
+	}
+	if p.Expiry == 0 {
+		return trace.BadParameter("expiry required")
+	}
+
+	return nil
+}
+
+// Sign will return a signed JSON token with the passed in claims embedded
+// within the token.
+func (k *Key) Sign(p *SignParams) (string, error) {
+	if k.privateKey == nil {
+		return "", trace.BadParameter("can not sign token with non-signing key")
+	}
+	if err := p.Check(); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	signingKey := jose.SigningKey{
+		Algorithm: jose.RS256,
+		Key:       k.privateKey,
+	}
+
+	sig, err := jose.NewSigner(signingKey, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	claims := jwtClaims{
+		Claims: josejwt.Claims{
```

Comment:
should the audience be set to application name?

Reply:
Is audience required?

Reply:
not strictly required, but good practice to prevent the confused deputy problem (token for app A shouldn't be valid for app B)